app/src/main/java/io/neurolab/main/TestModeActivity.java

```diff
@@ -0,0 +1,158 @@
+    private int baudRate = 9600;
+
+    private final String ACTION_USB_PERMISSION = "io.neurolab.main.USB_PERMISSION";
+
+    UsbSerialInterface.UsbReadCallback callBack = new UsbSerialInterface.UsbReadCallback() { //Defining a Callback which triggers whenever data is read.
+        @Override
+        public void onReceivedData(byte[] arg0) {
+            String data = null;
+            try {
+                data = new String(arg0, "UTF-8");
+                data.concat("/n");
+                setReadText(displayView, data);
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
+            }
+        }
+    };
+
+    private final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() { //Broadcast Receiver to automatically start and stop the Serial connection.
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            switch (intent.getAction()) {
+                case UsbManager.ACTION_USB_DEVICE_ATTACHED:
+                    deviceConnector.checkConnection(context);
+                    break;
+                case ACTION_USB_PERMISSION:
+                    boolean granted =
+                            intent.getExtras().getBoolean(UsbManager.EXTRA_PERMISSION_GRANTED);
+                    if (granted) {
+                        connection = usbManager.openDevice(device);
+                        serialPort = UsbSerialDevice.createUsbSerialDevice(device, connection);
+                        if (serialPort != null) {
+                            if (serialPort.open()) { //Set Serial Connection Parameters.
+                                writeBtn.setEnabled(true);
+                                serialPort.setBaudRate(baudRate);
+                                serialPort.setDataBits(UsbSerialInterface.DATA_BITS_8);
+                                serialPort.setStopBits(UsbSerialInterface.STOP_BITS_1);
+                                serialPort.setParity(UsbSerialInterface.PARITY_NONE);
+                                serialPort.setFlowControl(UsbSerialInterface.FLOW_CONTROL_OFF);
+                                serialPort.read(callBack);
+                                setReadText(displayView, getResources()
+                                        .getString(R.string.connection_opened) + "\n");
+                            } else {
+                                Log.d("SERIAL", "PORT NOT OPEN");
+                            }
+                        } else {
+                            Log.d("SERIAL", "PORT IS NULL");
+                        }
+                    } else {
+                        Log.d("SERIAL", "PERM NOT GRANTED");
+                    }
+                    break;
+                default:
+                    deviceConnector.checkConnection(context);
+                    break;
+            }
+        }
+    };
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
```

Comment:
It would be better if you add line breaks in the middle along with proper comments for better readability in your onCreate ( ) method. I mean it is looking very overwhelming at sight.

Reply:
Also since its an external library that you are using throughout for establishing a serial connection, please add comments at those function/variables usages so that it becomes easier to understand at a glance without digging in deeper for every contributor.